include/albatross/src/models/sparse_gp.hpp

```diff
@@ -315,23 +321,100 @@ class SparseGaussianProcessRegression
+        inducing_point_strategy_(this->covariance_function_, features);
+
+    const auto sc = sparse_components(features, u, targets);
+
+    const Eigen::Index m = sc.K_uu_llt.rows();
+    const Eigen::MatrixXd L_uu_inv =
+        sc.K_uu_llt.matrixL().solve(Eigen::MatrixXd::Identity(m, m));
+    const Eigen::MatrixXd BiLi = sc.B_ldlt.solve(L_uu_inv);
+    const Eigen::MatrixXd RtRBiLi = sc.RtR * sc.B_ldlt.solve(L_uu_inv);
+    const auto LT = sc.K_uu_llt.matrixL().transpose();
+    Eigen::VectorXd v = BiLi.transpose() * sc.P * sc.A_llt.solve(sc.y);
+
+    const Eigen::MatrixXd C_inv = LT.solve(RtRBiLi);
     DirectInverse solver(C_inv);
 
     using InducingPointFeatureType = typename std::decay<decltype(u[0])>::type;
     return Fit<GPFit<DirectInverse, InducingPointFeatureType>>(u, solver, v);
   }
 
+  template <typename FeatureType>
+  double log_likelihood(const RegressionDataset<FeatureType> &dataset) const {
+
+    const auto u =
+        inducing_point_strategy_(this->covariance_function_, dataset.features);
+
+    const SparseGPComponents sc =
+        sparse_components(dataset.features, u, dataset.targets);
+
+    // The log likelihood for y ~ N(0, K) is:
+    //
+    //   L = 1/2 (n log(2 pi) + log(|K|) + y^T K^-1 y)
+    //
+    // where in our case we have K = A + Q_ff
+
+    // First we get the determinant, |K|:
+    //
+    //   |K| = |A + Q_ff|
+    //       = |K_uu + K_uf A^-1 K_fu| |K_uu^-1| |A|
+    //       = |LL^T + K_uf A^-1 K_fu| |L^-T L^-1| |A|
+    //       = |L^-1| |LL^T + K_uf A^-1 K_fu| |L^-T| |A|
+    //       = |I + L^-1 K_uf A^-1 K_fu L^-T| |A|
+    //       = |I + P A^-1 P^T| |A|
+    //       = |B| |A|
+    //
+    // which is derived here (though there are errors in their derivation):
+    //   https://bwengals.github.io/pymc3-fitcvfe-implementation-notes.html
+
+    const double log_det_a = sc.A_llt.log_determinant();
+    const double log_det_b = sc.B_ldlt.vectorD().array().log().sum();
+    const double log_det = log_det_a + log_det_b;
+
+    // Then we need the Mahalanobis distance. To do so we'll draw heavily
+    // on some of the definitions used early in this module.
+    //
+    //   d = y^T K^-1 y
+    //     = y^T (A + Q_ff)^-1 y
+    //     = y^T (A^-1 + A^-1 K_fu S^-1 K_uf A^-1) y
+    //     = y^T (A^-1 + A^-1 P^T B^-1 P A^-1) y
+    //     = y^T A^-1 y + y^T A^-1 P^T B^-1 P A^-1 y
+    //     = y^T (A^-1 y) + (B^{-1/2} P A^-1 y)^T (B^{-1/2} P A^-1 y)
+    //     = y^T y_a + c^T c
+    //
+    // with
+    //
+    //   y_a = A^-1 y   and  c = B^{-1/2} P y_a
+
+    Eigen::VectorXd y_a = sc.A_llt.solve(sc.y);
+    Eigen::VectorXd c = sc.B_ldlt.sqrt_solve((sc.P * y_a).eval());
+
+    double log_quadratic = sc.y.transpose() * y_a;
+    log_quadratic -= c.transpose() * c;
+
+    const double rank = static_cast<double>(sc.y.size());
+    const double log_dimension = rank * log(2 * M_PI);
+
+    return -0.5 * (log_det + log_quadratic + log_dimension) +
+           this->prior_log_likelihood();
```

Comment:
Can you point to the formula used here?

Reply:
It's actually given in a comment at the top of the method.